### PR TITLE
Add aten::quantize_per_tensor and aten::int_repr

### PR DESCRIPTION
**Pull request by mgehre-amd**
This PR adds support for `quantize_per_tensor` and `int_repr` and their lowering to Linalg.
The use case is to support the pattern `int_repr(quantize_per_tensor(x, scale, zero_point, type))`, which is used by the brevitas quantizer.

While working on this PR, I discovered two issues that make lowering `quantize_per_tensor` fail for `quint8` and `qint32` (`qint8` works):
(1) For uint8, I get `ERROR: dtype (torch.int8) is not equal to golden dtype (torch.uint8)` in the e2e test. It's unclear to me why this happens and how to fix. After learning more about it, I will fix it in a follow-up PR.
(2) qint32 doesn't even have a type representation in `TorchTypes.td` yet.

The strategy to lower `quantize_per_tensor` is to do `cast(clamp(round(input/scale + zero_point))` into an integer type.
Doing that, `int_repr` becomes a no-op. If we would want to implement other operators that work on quantized types like qint8,
we would also need to carry the static information (i.e. scale and zero_point) to them. For `int_repr`, this is not needed because it doesn't use that information.

*Details*:
With this PR from
```
  def forward(self, x):
   return torch.int_repr(torch.quantize_per_tensor(x, 0.5, 0, torch.qint8))
```
we first get
```
module attributes {torch.debug_module_name = "TestModule"} {
  func.func @forward(%arg0: !torch.vtensor<[3],f32>, %arg1: !torch.vtensor<[3],f32>) -> !torch.vtensor<[3],si8> {
    %float5.000000e-01 = torch.constant.float 5.000000e-01
    %int0 = torch.constant.int 0
    %int12 = torch.constant.int 12
    %0 = torch.aten.quantize_per_tensor %arg0, %float5.000000e-01, %int0, %int12 : !torch.vtensor<[3],f32>, !torch.float, !torch.int, !torch.int -> !torch.vtensor<[3],!torch.qint8>
    %1 = torch.aten.int_repr %0 : !torch.vtensor<[3],!torch.qint8> -> !torch.vtensor<[3],si8>
    return %1 : !torch.vtensor<[3],si8>
  }
}
```
and then lower that into
```
 #map = affine_map<(d0) -> (d0)>
module attributes {torch.debug_module_name = "TestModule"} {
  ml_program.global private mutable @global_seed(dense<0> : tensor<i64>) : tensor<i64>
  func.func @forward(%arg0: tensor<3xf32>, %arg1: tensor<3xf32>) -> tensor<3xi8> {
    %cst = arith.constant 0.000000e+00 : f32
    %cst_0 = arith.constant 2.550000e+02 : f32
    %0 = tensor.empty() : tensor<3xi8>
    %1 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel"]} ins(%arg0 : tensor<3xf32>) outs(%0 : tensor<3xi8>) {
    ^bb0(%in: f32, %out: i8):
      %2 = arith.addf %in, %cst : f32
      %3 = math.roundeven %2 : f32
      %4 = arith.maxf %3, %cst : f32
      %5 = arith.minf %4, %cst_0 : f32
      %6 = arith.fptoui %5 : f32 to i8
      linalg.yield %6 : i8
    } -> tensor<3xi8>
    return %1 : tensor<3xi8>
  }
}
```

### CODE REVIEW

File: python/torch_mlir_e2e_test/test_suite/quantize.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    @export
+    @annotate_args([
+        None,
+        ([-1, -1], torch.float32, True),
+    ])
+    def forward(self, x):
+        y1 = torch.int_repr(torch.quantize_per_tensor(x, 1.0, 0, torch.qint8))
+        # FIXME: When lowering to TOSA, the type gets converted to int8
+        # due to convertDtypeToBuiltinElementType loosing signedness.
```
Comment (ljfitz) [inline python/torch_mlir_e2e_test/test_suite/quantize.py]:
```suggestion
        # due to convertDtypeToBuiltinElementType losing signedness.
```

File: python/torch_mlir_e2e_test/test_suite/quantize.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        # This causes
+        #   ERROR: dtype (torch.int8) is not equal to golden dtype (torch.uint8)
+        # y2 = torch.int_repr(torch.quantize_per_tensor(x, 1.0, 0, torch.quint8))
+        return y1
+
+
+@register_test_case(module_factory=lambda: Quantize())
+def Quantize_basic(module, tu: TestUtils):
+    module.forward(tu.rand(3, 5, low=-10, high=10))
+    print(module.forward(tu.rand(3, 5, low=-10, high=10)))
```
Comment (ljfitz) [inline python/torch_mlir_e2e_test/test_suite/quantize.py]: Residual debugging?

File: python/torch_mlir_e2e_test/test_suite/quantize.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        super().__init__()
+
+    @export
+    @annotate_args([
+        None,
+        ([-1, -1], torch.float32, True),
+    ])
+    def forward(self, x):
+        y1 = torch.int_repr(torch.quantize_per_tensor(x, 1.0, 0, torch.qint8))
+        # FIXME: When lowering to TOSA, the type gets converted to int8
```
Comment (ljfitz) [inline python/torch_mlir_e2e_test/test_suite/quantize.py]: This is to be fixed before upstreaming?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 4]
```diff
@@ -3093,6 +3093,58 @@ LogicalResult ConvertAtenOp<AtenBroadcastToOp>::matchAndRewrite(
...
 
+template <>
+LogicalResult ConvertAtenOp<AtenQuantizePerTensorOp>::matchAndRewrite(
+    AtenQuantizePerTensorOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+
+  auto resultTy = getTypeConverter()
+                      ->convertType(op->getOperand(0).getType());
+
+  // This is dead and will be removed
```
Comment (ljfitz) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: I'm guessing that this has to do with the pattern supported in `AtenIntReprOp`? I think cookie-trail comments are needed to help people understand and maintain the approach.

File: python/torch_mlir_e2e_test/test_suite/quantize.py (hunk 3 above)
Comment (mgehre-amd) [reply python/torch_mlir_e2e_test/test_suite/quantize.py]: I'm unsure. I don't want to make this PR to big, and having qint8 work is already a nice improvement.
What do you think?

File: python/torch_mlir_e2e_test/test_suite/quantize.py (hunk 2 above)
Comment (mgehre-amd) [reply python/torch_mlir_e2e_test/test_suite/quantize.py]: Oops, soory

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 4 above)
Comment (mgehre-amd) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Yes, I can explain this better. It feels a bit hacky, but essentially:
In `ConvertAtenOp<AtenIntReprOp>`, we replace
```
a = "torch.constant.int"() {value = 0 : i64} : () -> !torch.int
y = AtenQuantizePerTensorOp(x, offset=a)
z = AtenIntReprOp(y)
```
by 
```
a = "torch.constant.int"() {value = 0 : i64} : () -> !torch.int
y = AtenQuantizePerTensorOp(x, offset=a)
z = tosa.cast(x) // note the 'x' here
```
Now a, and y are dead. They are not removed during the TorchToTosa pass, and then it fails with
`error: failed to legalize operation 'torch.constant.int'`.

To work around this, I replace `AtenQuantizePerTensorOp` by tosa.cast, which magically gets rid of `a`,
and later this dead cast is removed by some other pass.

Comment (mgehre-amd) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Ok, I rewrote that part to instead:
1) replace AtenQuantizePerTensorOp by cast to int8
2) replace AtenIntReprOp by its operand (i.e. make it no-op)
This is much cleaner now.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 5]
```diff
@@ -3093,6 +3093,57 @@ LogicalResult ConvertAtenOp<AtenBroadcastToOp>::matchAndRewrite(
       "unimplemented: broadcasts other than same rank or zero ranked tensor.");
 }
 
+template <>
+LogicalResult ConvertAtenOp<AtenQuantizePerTensorOp>::matchAndRewrite(
+    AtenQuantizePerTensorOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+
```
Comment (mgehre-amd) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Comment for strategy of going to quantized types with TOSA

Comment (mgehre-amd) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Implement adds and muls

File: python/torch_mlir_e2e_test/test_suite/quantize.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
```
Comment (mgehre-amd) [inline python/torch_mlir_e2e_test/test_suite/quantize.py]: Mark XFAIL

**Comment by mgehre-amd** [comment]
@ljfitz,
- I added a description to this PR
- removed the TOSA lowering due to the rounding issue
- implemented a linalg lowering
- in linalg, I get further with the quint8, but I'm still stuck due to `ERROR: dtype (torch.int8) is not equal to golden dtype (torch.uint8)`. I added that issue into the PR description. I would like to still go ahead and merge this to at least have the qint8 case working.

What do you think?

File: e2e_testing/xfail_sets.py [hunk 7]
```diff
@@ -12,7 +12,9 @@
 
 from torch_mlir_e2e_test.test_suite import COMMON_TORCH_MLIR_LOWERING_XFAILS
 
-LINALG_XFAIL_SET = COMMON_TORCH_MLIR_LOWERING_XFAILS
+LINALG_XFAIL_SET = COMMON_TORCH_MLIR_LOWERING_XFAILS.union({
+    "Quantize_uint8"
```
Comment (ljfitz) [inline e2e_testing/xfail_sets.py:17]: In your description you suggest you want help understanding why this fails. I suggest you also add that in a comment here.

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 8]
```diff
@@ -907,7 +923,8 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
                      .getElementType();
 
     Value self = payloadArgs[0];
-    Value threshold = convertScalarToDtype(b, loc, adaptor.getThreshold(), dtype);
+    Value threshold =
```
Comment (ljfitz) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp]: This looks like an unrelated change - please see the 2nd last paragraph of https://llvm.org/docs/CodingStandards.html#introduction.

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 9]
```diff
@@ -1506,7 +1571,8 @@ class ConvertAtenDetachOp : public OpConversionPattern<AtenDetachOp> {
       return failure();
 
     Type resultType = getTypeConverter()->convertType(op.getType());
-    rewriter.replaceOpWithNewOp<tensor::CastOp>(op, resultType, adaptor.getSelf());
+    rewriter.replaceOpWithNewOp<tensor::CastOp>(op, resultType,
```
Comment (ljfitz) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp]: Unrelated reformatting?

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 10]
```diff
@@ -929,7 +946,8 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
 
     Value grad = convertScalarToDtype(b, loc, payloadArgs[0], dtype);
     Value self = convertScalarToDtype(b, loc, payloadArgs[1], dtype);
-    Value threshold = convertScalarToDtype(b, loc, adaptor.getThreshold(), dtype);
+    Value threshold =
```
Comment (ljfitz) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp]: Unrelated reformatting?

**Review by ljfitz — APPROVED** [review]
I'd be happy for you to use this to kick off the discussion upstream.

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp (hunk 9 above)
Comment (mgehre-amd) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp]: Hm, my "format changed lines" in vscode doesn't seem to always work. Sorry for missing that before review.
